### PR TITLE
Add Status Query endpoint support

### PR DESCRIPTION
**Pull request by jujaga**
# Description

This PR implements the /status queryable endpoint as specified in the specification. 
- Adds missing 422 validation support to the /status endpoint family
- Improve unit test code structure for mocking classes
- Refactor transformer into a component
- Refactor validators into model and validator functions

[[SHOWCASE-326]](https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-326)

## Types of changes

New feature (non-breaking change which adds functionality)
Documentation (non-breaking change with enhancements to documentation)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the OpenAPI 3.0 `v*.api-spec.yaml` documentation (if appropriate)
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-9f0fbe-tools.pathfinder.gov.bc.ca/dashboard?id=common-hosted-email-service-pr-26

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-9f0fbe-tools.pathfinder.gov.bc.ca/dashboard?id=common-hosted-email-service-pr-26

File: app/tests/unit/routes/v1/email.spec.js [hunk 1]
```diff
@@ -35,6 +26,10 @@ describe(`POST ${basePath}`, () => {
   });
 
   it('should push a message and yield an Ethereal url', async () => {
+    ChesService.prototype.sendEmail.mockImplementation(() => {
```
Comment (parc-jason) [inline app/tests/unit/routes/v1/email.spec.js]: Since the test is on the route, you can't use this mock implementation.  You will not be testing if the route is indicating whether or not to invoke ethereal. The test should be to test if devMode=true will fire ethereal, but this always returns an ethereal url from the chesSvc.  Get this test to actually test devMode=true, and another one that tests the negative case.

File: app/tests/unit/routes/v1/merge.spec.js [hunk 2]
```diff
@@ -72,6 +63,7 @@ describe(`POST ${basePath}`, () => {
   });
 
   it('should push a message and yield an Ethereal url', async () => {
+    ChesService.prototype.sendEmailMerge.mockResolvedValue('https://example.com');
```
Comment (parc-jason) [inline app/tests/unit/routes/v1/merge.spec.js]: Same as the email comment, the mock service call needs to handle ethereal true/false, so we can actually test the route is doing the work figuring out if we are in devMode.  Have one with devMode=true and another with devMode spelled incorrectly and it should return a "normal" response.

File: app/tests/unit/components/transformer.spec.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
+const helper = require('../../common/helper');
+const transformer = require('../../../src/components/transformer');
+
+helper.logHelper();
+
+describe('toStatusResponse', () => {
+  it('should return an empty object if all properties are undefined', () => {
```
Comment (parc-jason) [inline app/tests/unit/components/transformer.spec.js]: Need test for undefined, and "bad" object types (string, number, array)

File: app/tests/unit/components/transformer.spec.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    expect(result).toBeTruthy();
+    expect(Object.keys(result).length).toBe(1);
+    expect(result.updatedTimestamp).toBe(1571679721833);
+  });
+});
+
+describe('toTransactionResponse', () => {
+  it('should return an object with all properties populated', () => {
+    const result = transformer.toTransactionResponse({
```
Comment (parc-jason) [inline app/tests/unit/components/transformer.spec.js]: Need test for undefined, and "bad" object types (string, number, array, structure of "bad" objects)

File: app/tests/unit/components/transformer.spec.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
+const helper = require('../../common/helper');
+const transformer = require('../../../src/components/transformer');
+
+helper.logHelper();
+
+describe('toStatusResponse', () => {
+  it('should return an empty object if all properties are undefined', () => {
+    const result = transformer.toStatusResponse({});
```
Comment (parc-jason) [inline app/tests/unit/components/transformer.spec.js]: See other comments about always returning all object keys...

File: app/tests/unit/routes/v1/status.spec.js [hunk 6]
```diff
@@ -4,66 +4,145 @@ const request = require('supertest');
...
+      description: null,
+      status: 'accepted',
+      timestamp: 1571679721833
+    }
+  ];
+
+  return status;
+};
+
+describe(`POST ${basePath}`, () => {
```
Comment (parc-jason) [inline app/tests/unit/routes/v1/status.spec.js]: Think it should be GET ${basePath}

File: app/tests/unit/routes/v1/status.spec.js [hunk 7]
```diff
@@ -4,66 +4,145 @@ const request = require('supertest');
...
+    expect(response.body.details).toMatch(errorMsg);
+  });
+
+  it('should respond with a validation error', async () => {
+    const response = await request(app).get(`${basePath}`);
+
+    expect(response.statusCode).toBe(422);
+    expect(response.body).toBeTruthy();
+    expect(response.body.detail).toMatch('Validation failed');
+    expect(response.body.errors).toHaveLength(1);
```
Comment (parc-jason) [inline app/tests/unit/routes/v1/status.spec.js]: Should we test which validation error is actually returned?

File: app/tests/unit/routes/v1/status.spec.js [hunk 8]
```diff
@@ -79,9 +158,24 @@ describe(`POST ${basePath}/:msgId`, () => {
...
   });
 
+  it('should respond with a validation error', async () => {
+    const id = 'badId';
+    const response = await request(app).get(`${basePath}/${id}`);
+
+    expect(response.statusCode).toBe(422);
+    expect(response.body).toBeTruthy();
+    expect(response.body.detail).toMatch('Validation failed');
+    expect(response.body.errors).toHaveLength(1);
```
Comment (parc-jason) [inline app/tests/unit/routes/v1/status.spec.js]: Test the actual validation error returned.

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-9f0fbe-tools.pathfinder.gov.bc.ca/dashboard?id=common-hosted-email-service-pr-26

File: app/tests/unit/routes/v1/status.spec.js (hunk 6 above)
Comment (jujaga) [reply app/tests/unit/routes/v1/status.spec.js]: Fixed

File: app/tests/unit/components/transformer.spec.js (hunk 5 above)
Comment (jujaga) [reply app/tests/unit/components/transformer.spec.js]: Implemented

File: app/tests/unit/routes/v1/status.spec.js (hunk 8 above)
Comment (jujaga) [reply app/tests/unit/routes/v1/status.spec.js]: Implemented

File: app/tests/unit/routes/v1/status.spec.js (hunk 7 above)
Comment (jujaga) [reply app/tests/unit/routes/v1/status.spec.js]: Not extremely necessary here but is good to do to be consistent. Implemented.

File: app/tests/unit/components/transformer.spec.js (hunk 3 above)
Comment (jujaga) [reply app/tests/unit/components/transformer.spec.js]: Implemented

File: app/tests/unit/components/transformer.spec.js (hunk 4 above)
Comment (jujaga) [reply app/tests/unit/components/transformer.spec.js]: Implemented

File: app/src/routes/v1/merge.js [hunk 9]
```diff
@@ -2,7 +2,6 @@ const mergeComponent = require('../../components/merge');
 
 const mergeRouter = require('express').Router();
 const { validateMerge } = require('../../middleware/validation');
-
```
[the remark this answers is not in the document]
Comment (jujaga) [reply app/src/routes/v1/merge.js]: Updated - thanks for the catch

File: app/src/docs/v1.api-spec.yaml [hunk 10]
```diff
@@ -179,8 +179,9 @@ paths:
       description: >-
         This is a freeform query endpoint, and reasonably expects at least one
         of the parameters to be populated. To minimize response size, only
-        msgId, status, tag and txId properties will appear if no other fields
-        are explicitly requested.
+        msgId, status, tag and txId properties appear by default. Properties
```
[the remark this answers is not in the document]
Comment (jujaga) [reply app/src/docs/v1.api-spec.yaml]: Updated

File: app/src/components/transformer.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+          status: h.status,
+          timestamp: moment.utc(h.createdAt).valueOf()
+        };
+      }) : undefined,
+      tag: msg.tag,
+      txId: msg.transactionId,
+      updatedTimestamp: msg.updatedAt ? moment.utc(msg.updatedAt).valueOf() : undefined
+    };
+
+    return utils.dropNullAndUndefinedObject(result);
```
[the remark this answers is not in the document]
Comment (jujaga) [reply app/src/components/transformer.js]: Updated to align with proposal.

File: app/src/services/chesSvc.js [hunk 12]
```diff
@@ -40,49 +40,98 @@ class ChesService {
...
+   *  @param {string} status - the desired status of the messages
+   *  @param {string} tag - the desired tag of the messages
+   *  @param {string} transactionId - the id of the desired transaction
+   *  @param {string} fields - a comma separated list of desired columns in addition to the defaults
+   *  @throws Problem if an unexpected error occurs
+   *  @returns {object[]} Array of Status objects with a subset of properties
+   */
+  async findStatuses(client, messageId, status, tag, transactionId, fields) {
+    let fieldArray;
+    if (fields) {
```
[the remark this answers is not in the document]
Comment (jujaga) [reply app/src/services/chesSvc.js]: The transformer should house all functional/static object shape conversions. While the fields could be considered a shape change in their own right, that block is only really renaming attributes of an in-transit object. Since the source and destination objects do not actually represent API nor DB Model objects, I don't think it should be in the transformer.

This though is a completely null point however as we decided to not use the fields parameter anymore anyways.

File: app/tests/unit/routes/v1/email.spec.js (hunk 1 above)
Comment (jujaga) [reply app/tests/unit/routes/v1/email.spec.js]: We can actually satisfy checking whether the sendEmail function was invoked to run with ethereal as true or false by employing a jest spy. Assigning the spy to the mock, we can check what arguments were passed into the mocked function.
`expect(spy).toHaveBeenCalledWith(undefined, body, true);`

File: app/tests/unit/routes/v1/merge.spec.js (hunk 2 above)
Comment (jujaga) [reply app/tests/unit/routes/v1/merge.spec.js]: We can actually satisfy checking whether the sendEmailMerge function was invoked to run with ethereal as true or false by employing a jest spy. Assigning the spy to the mock, we can check what arguments were passed into the mocked function.
expect(spy).toHaveBeenCalledWith(undefined, body, true);
